### DISCUSSION TITLE
[Question] how to download file with FastAPI

### DISCUSSION THREAD

**Discussion by iromame** [flet-dev/flet#3632] [Q&A]
### Question

I have a question about downloading files when executed in a web browser.

https://flet.dev/docs/controls/filepicker/
This URL show code to download the file when executed in a web browser, though,
The code is incomplete and an error occurs when I run it as is.

I did my own research and wrote the code.
When I ran this code with uvicorn, it worked.

$ uvicorn my-app:app --reload

I am using flet 0.23.2, windows10

file tree is following

C:.
│  my-app.py
│
├─files
│      samplefile.zip

However, I am not familiar with web apps and do not understand this code very well and am not sure

* Please let me know if there is a better code
* Why doesn't app = ft.app(main, export_asgi_app=True) or app = flet_fastapi.app(main) work?
* Can I run it with python my-app.py or flet run my-app.py?

I am confused by the frequent updates.
It would be very helpful if you could tell me.

### Code sample

```python
from fastapi.responses import FileResponse
import flet.fastapi as flet_fastapi
import flet as ft

def main(page: ft.Page):
    page.add(ft.ElevatedButton("Download myfile", on_click=lambda _: page.launch_url("/download/samplefile.zip")))


app = flet_fastapi.FastAPI()

"""
following code not worked
#app = ft.app(main, export_asgi_app=True)
#app = flet_fastapi.app(main)
"""

@app.get("/download/{filename}")
def download(filename: str):
    path = f"files/{filename}"
    return FileResponse(path)

app.mount("/", flet_fastapi.app(main))
```

### ------------------------------------------------------

- [X] I have searched for answers to my question both in the issues and in previous discussions.

**Comment by fuad00** [comment]
any updates on this?

**Comment by tanmay-bhatgare** [comment]
After reviewing your code, problem seems to be the way your code work. See major mistake is you're not providing filename as attribute to `FileResponse` which causing `FileResponse` to return None. Second is how you run your app, according to `flet_fastapi` docs, you have to run your app using uvicorn, not by `python` or `flet` cli.

## here's how you can achieve desired functionality.
#### Structure
```bash
test_dir
│
├───test.py               # Your Flet FastAPI app
├───files                 # Directory containing files to download
│     └───sample.txt      # Sample file for download
└───__pycache__           # Will be generated automatically by Python
```
#### Code
```python
from fastapi.responses import FileResponse
import flet.fastapi as flet_fastapi
import flet as ft

def main(page: ft.Page):
    page.add(ft.ElevatedButton("Download myfile", on_click=lambda _: page.launch_url("/download/sample.txt")))


app = flet_fastapi.FastAPI()

@app.get("/download/{filename}")
def download(filename: str):
    path = f"files/{filename}"
    print(path)
    return FileResponse(path=path, filename=filename, media_type="text/plain") # provide filename, media_type is optional

app.mount("/", flet_fastapi.app(main))
```
#### Command to run
```bash
# inside test_dir
uvicorn test:app --reload
```
